New Feature: Draggable Track

felix-kaestner opened the pull request:
This Pull-Request adds the functionality to drag the track of the range slider, moving all values at the same time, keeping the difference between values the same. This API is exposed via a new prop `draggableTrack` which is only used, when there are multiple values. Here is a quick demo:

![DraggableTrack](https://user-images.githubusercontent.com/23213965/93664954-fb722900-fa72-11ea-90c8-207cd2463ee2.gif)

I added example stories to showcase the api with two or multiple thumbs, overlapping thumbs, labels and also RTL.

Would be happy to discuss changes if you have any recommendations on what yould be improved!

tajo commented on src/Range.tsx:
```diff
@@ -209,17 +211,41 @@ class Range extends React.Component<IProps> {
-      () => this.onMove(e.clientX, e.clientY)
-    );
+    if (this.props.values.length > 1) {
+      // in case a thumb was clicked, move the thumb
```
this comment applies to the `else` bellow?

tajo commented:
Thanks, I like this feature. Some notes:

- There should be an example for the vertical setup. On the other hand, examples `AllowOverlapDraggableTrack` and `LabeledTwoThumbsDraggableTrack` don't really demo anything interesting, so maybe just removing them?
- There should be an e2e test for this feature, ideally for both vertical and horizontal movement.
- It breaks the marks feature and examples. 
- Flow types should be updated.
- README should be updated.

tajo commented:
On the second thought, how would you describe use-case for this? How did you come up with this? It's not a known pattern afaik.

Also, there is a valid request to support `click-track-to-move-the-thumb` for multiple thumbs scenario which seems like a common pattern across other libraries and it would somewhat collide with this feature.

felix-kaestner commented:
@tajo Thanks for your feedback. I would like to address the points you made and improve on this PR in the upcoming week.

> There should be an example for the vertical setup. On the other hand, examples AllowOverlapDraggableTrack and LabeledTwoThumbsDraggableTrack don't really demo anything interesting, so maybe just removing them?

Seems reasonable. So I will add a vertical example and remove the ones you mentioned. 👍 

> On the second thought, how would you describe use-case for this? 

The use-case for us, is that we are using it in a large-scale visualization app, that we are building on top off [deck.gl](https://github.com/visgl/deck.gl) and React. In such application you typically have a lot of different filters, which you can use to adjust the visualization. With this feature you would be able to e.g. select a time-window that you can then drag over the duration of a day. [Here is a pretty good example.](https://movement.uber.com/explore/san_francisco/mobility-heatmap/query?lang=en-US)

> How did you come up with this? It's not a known pattern afaik.

I have to agree that this is not a "standard feature", but there are a some range/slider libraries that support such pattern, most noticeably [react-input-range](https://codepen.io/davidchin/full/GpNvqw/) (the last one) and the example mentioned above.

> Also, there is a valid request to support `click-track-to-move-the-thumb` for multiple thumbs scenario which seems like a common pattern across other libraries and it would somewhat collide with this feature.

Since the `draggableTrack` prop is _opt-in_ meaning that it is set to `false` per default, I don't think these features would collide, since a user would have to actively enable this feature. 
So if `draggableTrack` is set to `false` (default) one could use the `click-track-to-move-the-thumb`, while if the user chooses to set `draggableTrack` to `true` one would use the `click-track-to-move-everything` gesture.

felix-kaestner commented:
@tajo I addressed all of your points. Additionally, I implemented the `click-track-to-move-the-closest-thumb` gesture for sliders with multiple thumbs since it also simplifies the code. Would appreciate it if you could re-review this Pull Request.

tajo commented:
Looks go to me!

One last thing: Can you please remove all the image changes / additions from this PR? It's not setup very reliably so I'll need to create the new ones from my env to match the CI (and also want to see that the your changes don't impact existing e2e snapshots)

felix-kaestner commented:
@tajo Thanks. I removed all changes/additions in the `e2e/__image_snapshots__` directory.
